CORS-3571: permissions: add new create command and asset

r4f4 opened the pull request:
The new `create permissions` command and the accompanying `Permissions` asset generates a file with required permissions for a given `install-config.yaml` in supported platforms.

For AWS it means a policy json file with a list of `Allow` actions needed by the installer.

r4f4 commented:
Example of using the command for a bare minimum `install-config.yaml`:
```yaml
apiVersion: v1
metadata:
  name: rdossant-installer-07041559
publish: External
pullSecret: <redacted>
sshKey: <redacted>
baseDomain: devcluster.openshift.com
platform:
  aws:
    region: us-east-2
```
```bash
[root@7f3fe75ddd7c c] $ ./openshift-install create permissions --log-level debug
DEBUG OpenShift Installer unreleased-master-10507-gceac7228c8bf38a3df9892e4737c267bbaeab30c-dirty
DEBUG Built from commit ceac7228c8bf38a3df9892e4737c267bbaeab30c
DEBUG Fetching Permissions...
DEBUG Loading Permissions...
DEBUG   Loading Install Config...
DEBUG     Loading SSH Key...
DEBUG     Loading Base Domain...
DEBUG       Loading Platform...
DEBUG     Loading Cluster Name...
DEBUG       Loading Base Domain...
DEBUG       Loading Platform...
DEBUG     Loading Pull Secret...
DEBUG     Loading Platform...
DEBUG   Using Install Config loaded from target directory
DEBUG   Fetching Install Config...
DEBUG   Reusing previously-fetched Install Config
DEBUG Generating Permissions...
INFO Consuming Install Config from target directory
DEBUG Purging asset "Install Config" from disk
INFO Permissions created in: .
[root@7f3fe75ddd7c c] $ jq <aws-permissions-policy.json
```
```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "autoscaling:DescribeAutoScalingGroups",
        "ec2:AllocateAddress",
        "ec2:AssociateAddress",
        "ec2:AssociateDhcpOptions",
        "ec2:AssociateRouteTable",
        "ec2:AttachInternetGateway",
        "ec2:AttachNetworkInterface",
        "ec2:AuthorizeSecurityGroupEgress",
        "ec2:AuthorizeSecurityGroupIngress",
        "ec2:CopyImage",
        "ec2:CreateDhcpOptions",
        "ec2:CreateInternetGateway",
        "ec2:CreateNatGateway",
        "ec2:CreateNetworkInterface",
        "ec2:CreateRoute",
        "ec2:CreateRouteTable",
        "ec2:CreateSecurityGroup",
        "ec2:CreateSubnet",
        "ec2:CreateTags",
        "ec2:CreateVolume",
        "ec2:CreateVpc",
        "ec2:CreateVpcEndpoint",
        "ec2:DeleteDhcpOptions",
        "ec2:DeleteInternetGateway",
        "ec2:DeleteNatGateway",
        "ec2:DeleteNetworkInterface",
        "ec2:DeletePlacementGroup",
        "ec2:DeleteRoute",
        "ec2:DeleteRouteTable",
        "ec2:DeleteSecurityGroup",
        "ec2:DeleteSnapshot",
        "ec2:DeleteSubnet",
        "ec2:DeleteTags",
        "ec2:DeleteVolume",
        "ec2:DeleteVpc",
        "ec2:DeleteVpcEndpoints",
        "ec2:DeregisterImage",
        "ec2:DescribeAccountAttributes",
        "ec2:DescribeAddresses",
        "ec2:DescribeAvailabilityZones",
        "ec2:DescribeDhcpOptions",
        "ec2:DescribeImages",
        "ec2:DescribeInstanceAttribute",
        "ec2:DescribeInstanceCreditSpecifications",
        "ec2:DescribeInstances",
        "ec2:DescribeInternetGateways",
        "ec2:DescribeKeyPairs",
        "ec2:DescribeNatGateways",
        "ec2:DescribeNetworkAcls",
        "ec2:DescribeNetworkInterfaces",
        "ec2:DescribePrefixLists",
        "ec2:DescribeRegions",
        "ec2:DescribeRouteTables",
        "ec2:DescribeSecurityGroupRules",
        "ec2:DescribeSecurityGroups",
        "ec2:DescribeSubnets",
        "ec2:DescribeTags",
        "ec2:DescribeVolumes",
        "ec2:DescribeVpcAttribute",
        "ec2:DescribeVpcClassicLink",
        "ec2:DescribeVpcClassicLinkDnsSupport",
        "ec2:DescribeVpcEndpoints",
        "ec2:DescribeVpcs",
        "ec2:DetachInternetGateway",
        "ec2:DisassociateRouteTable",
        "ec2:GetEbsDefaultKmsKeyId",
        "ec2:ModifyInstanceAttribute",
        "ec2:ModifyNetworkInterfaceAttribute",
        "ec2:ModifySubnetAttribute",
        "ec2:ModifyVpcAttribute",
        "ec2:ReleaseAddress",
        "ec2:ReplaceRouteTableAssociation",
        "ec2:RevokeSecurityGroupEgress",
        "ec2:RevokeSecurityGroupIngress",
        "ec2:RunInstances",
        "ec2:TerminateInstances",
        "elasticloadbalancing:AddTags",
        "elasticloadbalancing:ApplySecurityGroupsToLoadBalancer",
        "elasticloadbalancing:AttachLoadBalancerToSubnets",
        "elasticloadbalancing:ConfigureHealthCheck",
        "elasticloadbalancing:CreateListener",
        "elasticloadbalancing:CreateLoadBalancer",
        "elasticloadbalancing:CreateLoadBalancerListeners",
        "elasticloadbalancing:CreateTargetGroup",
        "elasticloadbalancing:DeleteLoadBalancer",
        "elasticloadbalancing:DeleteTargetGroup",
        "elasticloadbalancing:DeregisterInstancesFromLoadBalancer",
        "elasticloadbalancing:DeregisterTargets",
        "elasticloadbalancing:DescribeInstanceHealth",
        "elasticloadbalancing:DescribeListeners",
        "elasticloadbalancing:DescribeLoadBalancerAttributes",
        "elasticloadbalancing:DescribeLoadBalancers",
        "elasticloadbalancing:DescribeTags",
        "elasticloadbalancing:DescribeTargetGroupAttributes",
        "elasticloadbalancing:DescribeTargetGroups",
        "elasticloadbalancing:DescribeTargetHealth",
        "elasticloadbalancing:ModifyLoadBalancerAttributes",
        "elasticloadbalancing:ModifyTargetGroup",
        "elasticloadbalancing:ModifyTargetGroupAttributes",
        "elasticloadbalancing:RegisterInstancesWithLoadBalancer",
        "elasticloadbalancing:RegisterTargets",
        "elasticloadbalancing:SetLoadBalancerPoliciesOfListener",
        "elasticloadbalancing:SetSecurityGroups",
        "iam:AddRoleToInstanceProfile",
        "iam:CreateInstanceProfile",
        "iam:CreateRole",
        "iam:DeleteAccessKey",
        "iam:DeleteInstanceProfile",
        "iam:DeleteRole",
        "iam:DeleteRolePolicy",
        "iam:DeleteUser",
        "iam:GetInstanceProfile",
        "iam:GetRole",
        "iam:GetRolePolicy",
        "iam:GetUser",
        "iam:ListAttachedRolePolicies",
        "iam:ListInstanceProfiles",
        "iam:ListInstanceProfilesForRole",
        "iam:ListRolePolicies",
        "iam:ListRoles",
        "iam:ListUserPolicies",
        "iam:ListUsers",
        "iam:PassRole",
        "iam:PutRolePolicy",
        "iam:RemoveRoleFromInstanceProfile",
        "iam:SimulatePrincipalPolicy",
        "iam:TagInstanceProfile",
        "iam:TagRole",
        "route53:ChangeResourceRecordSets",
        "route53:ChangeTagsForResource",
        "route53:CreateHostedZone",
        "route53:DeleteHostedZone",
        "route53:GetChange",
        "route53:GetHostedZone",
        "route53:ListHostedZones",
        "route53:ListHostedZonesByName",
        "route53:ListResourceRecordSets",
        "route53:ListTagsForResource",
        "route53:UpdateHostedZoneComment",
        "s3:CreateBucket",
        "s3:DeleteBucket",
        "s3:DeleteObject",
        "s3:GetAccelerateConfiguration",
        "s3:GetBucketAcl",
        "s3:GetBucketCors",
        "s3:GetBucketLocation",
        "s3:GetBucketLogging",
        "s3:GetBucketObjectLockConfiguration",
        "s3:GetBucketPolicy",
        "s3:GetBucketRequestPayment",
        "s3:GetBucketTagging",
        "s3:GetBucketVersioning",
        "s3:GetBucketWebsite",
        "s3:GetEncryptionConfiguration",
        "s3:GetLifecycleConfiguration",
        "s3:GetObject",
        "s3:GetObjectAcl",
        "s3:GetObjectTagging",
        "s3:GetObjectVersion",
        "s3:GetReplicationConfiguration",
        "s3:ListBucket",
        "s3:ListBucketVersions",
        "s3:PutBucketAcl",
        "s3:PutBucketPolicy",
        "s3:PutBucketTagging",
        "s3:PutEncryptionConfiguration",
        "s3:PutObject",
        "s3:PutObjectAcl",
        "s3:PutObjectTagging",
        "tag:GetResources"
      ],
      "Resource": [
        "*"
      ]
    }
  ]
}
```

r4f4 commented:
Update: rebased to fix merge conflicts and added CCO/cluster permissions to the list. With this, it should now be possible to create an user with the generated policy to deploy a cluster with only the required permissions.

r4f4 commented:
/retitle: WIP: CORS-3571: permissions: add new create command and asset

patrickdillon commented on pkg/asset/permissions/permissions.go:
```diff
@@ -44,35 +47,52 @@ func (o *Permissions) Generate(ctx context.Context, dependencies asset.Parents)
+		passGroups := append([]awsconfig.PermissionGroup{awsconfig.PermissionPassthroughCreds}, reqGroups...)
+		if err := o.writePolicy(passGroups, awsPolicyPassthroughFilename); err != nil {
+			return fmt.Errorf("failed to generate passthrough credentials permissions: %w", err)
 		}
```
we know from the install config which mode the CCO will run in, so should we just create the policy for the specified mode, rather than both?

r4f4 replied:
Interesting. If we know, then why do we do [this](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/permissions.go#L332-L351)?

r4f4 replied:
Would it be to cover the `ManualCredentials` case?

r4f4 replied:
Fixed.

patrickdillon reviewed (commented):
At a high level this all looks good to me. Left a couple of comments/suggestions.

r4f4 commented:
/retitle CORS-3571: permissions: add new create command and asset

r4f4 commented on cmd/openshift-install/create.go:
```diff
@@ -148,7 +148,16 @@ var (
+	permissionsTarget = target{
+		name: "Permissions",
+		command: &cobra.Command{
+			Use:   "permissions",
```
Done

r4f4 commented:
Update: addressed review comments, made the new command hidden so it doesn't show up with `--help` and made the code take credentials mode into consideration.

r4f4 commented:
/cc @mtulio

r4f4 commented:
`e2e-aws-ovn-imdsv2`: not affect by this change but something is wrong with the job. The [install-config.yaml](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_installer/8704/pull-ci-openshift-installer-master-e2e-aws-ovn-imdsv2/1845030349239750656/artifacts/e2e-aws-ovn-imdsv2/ipi-install-install/artifacts/install-config.yaml) has
```
controlPlane:
  platform:
    aws:
      zones:
      - us-east-1b
      - us-east-1c
      type: m6a.xlarge
      metadataService:
```
even though the [conf step](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_installer/8704/pull-ci-openshift-installer-master-e2e-aws-ovn-imdsv2/1845030349239750656/artifacts/e2e-aws-ovn-imdsv2/ipi-conf-aws/artifacts/install-config-metadata-auth.yaml.patch) has:
```
controlPlane:
  platform:
    aws:
      metadataService:
        authentication: Required
```

r4f4 commented on pkg/asset/installconfig/aws/permissions.go:
```diff
@@ -301,6 +307,79 @@ var permissions = map[PermissionGroup][]string{
+		"elasticloadbalancing:DescribeLoadBalancers",
+		"elasticloadbalancing:DescribeTargetGroups",
+		"elasticloadbalancing:RegisterTargets",
+		"ec2:DescribeVpcs",
```
This was taken verbatim from CCO: https://github.com/openshift/cloud-credential-operator/blob/master/pkg/aws/utils.go#L35-L92

r4f4 commented:
/test e2e-aws-ovn-min-perms

r4f4 commented:
Good news! The workflow+test works as the [aws-ovn-min-perms](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/8704/pull-ci-openshift-installer-master-e2e-aws-ovn-min-perms/1846460586150334464) job failed because we set the instance type in the install-config.yaml but the user doesn't have the `ec2:DescribeInstanceTypes` permission

r4f4 commented:
/test e2e-aws-ovn-min-perms

r4f4 commented:
/hold

I will break out the bug fixes into separate PRs.

r4f4 commented:
/test e2e-aws-ovn-min-perms

r4f4 commented:
/hold cancel

The permission issue is being addressed by https://github.com/openshift/installer/pull/9106. In the meantime, I think it's good that `aws-min-perms` is failing, since we can easily validate it's using the new command and the CI steps to provision the user.

mtulio commented on pkg/asset/permissions/permissions.go:
```diff
@@ -0,0 +1,105 @@
+
+	policy := iamv1.PolicyDocument{
+		Version: "2012-10-17",
+		Statement: []iamv1.StatementEntry{
+			{
+				Effect:   "Allow",
+				Action:   perms,
+				Resource: iamv1.Resources{"*"},
+			},
+		},
```
@r4f4 Is it possible to set the `Sid` by group? I wonder if we could adopt the approach of creating more verbose entries, generating an detailed self-explained policy, helping users to quickly identify reason of the permission with a descriptive SID.

For example, instead, generate a policy like:
```
		Statement: []iamv1.StatementEntry{
			{
				Effect:   "Allow",
				Action:   perms,   // content of PermissionCreateBase
				Resource: iamv1.Resources{"*"},
				Sid: "InstallerCreateBase",
			},
			{
				Effect:   "Allow",
				Action:   perms,   // content of PermissionDeleteBase
				Resource: iamv1.Resources{"*"},
				Sid: "InstallerDeleteBase",
			},
			{
				Effect:   "Allow",
				Action:   perms,     // content of PermissionMintCreds
				Resource: iamv1.Resources{"*"},
				Sid: "CcoMintBase",
			},
...
		},
```

It would be easier in the future if we can refine the resources/conditions for each group (best practice).

r4f4 replied:
@mtulio https://github.com/openshift/installer/pull/8704/commits/d2fc0ab411673243954c85575028b9d4a204a5de is the best I can do programmatically right now. Here is an excerpt:
```json
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Sid": "create-hosted-zone",
      "Effect": "Allow",
      "Action": [
        "route53:CreateHostedZone"
      ],
      "Resource": [
        "*"
      ]
    },
    {
      "Sid": "delete-base",
      "Effect": "Allow",
      "Action": [
        "autoscaling:DescribeAutoScalingGroups",
        "ec2:DeleteNetworkInterface",
        "ec2:DeletePlacementGroup",
        "ec2:DeleteTags",
        "ec2:DeleteVolume",
        "elasticloadbalancing:DeleteTargetGroup",
        "elasticloadbalancing:DescribeTargetGroups",
        "iam:DeleteAccessKey",
        "iam:DeleteUser",
        "iam:ListAttachedRolePolicies",
        "iam:ListInstanceProfiles",
        "iam:ListRolePolicies",
        "iam:ListUserPolicies",
        "s3:DeleteBucket",
        "s3:DeleteObject",
        "s3:ListBucketVersions",
        "tag:GetResources"
      ],
      "Resource": [
        "*"
      ]
    },
```

r4f4 replied:
@patrickdillon WDYT? ^

mtulio replied:
> @mtulio [d2fc0ab](https://github.com/openshift/installer/commit/d2fc0ab411673243954c85575028b9d4a204a5de) is the best I can do programmatically right now. Here is an excerpt:


@r4f4 LGTM, that's great.

r4f4 commented:
/test e2e-aws-ovn-min-perms

r4f4 commented:
/test e2e-aws-ovn-min-perms

r4f4 commented:
`e2e-aws-ovn-min-perms`: the [generated policy](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/test-platform-results/pr-logs/pull/openshift_installer/8704/pull-ci-openshift-installer-master-e2e-aws-ovn-min-perms/1848744291816443904/artifacts/e2e-aws-ovn-min-perms/ipi-conf-aws-user-min-permissions/artifacts/aws-permissions-policy-creds.json) and the min perms user were employed for the install and the expected error message:
```
 level=info msg=Credentials loaded from the "default" profile in file "/tmp/secret/aws_minimal_permission"
level=error msg=failed to fetch Master Machines: failed to load asset "Install Config": failed to create install config: [controlPlane.platform.aws: Internal error: error listing instance types: fetching instance types: UnauthorizedOperation: You are not authorized to perform this operation. User: arn:aws:iam::460538899914:user/ci-op-1m6j03lr-ccf9a-minimal-perm is not authorized to perform: ec2:DescribeInstanceTypes because no identity-based policy allows the ec2:DescribeInstanceTypes action
level=error msg=	status code: 403, request id: e07da609-55c6-4337-b821-d57b53b7c7cf, compute[0].platform.aws: Internal error: error listing instance types: fetching instance types: UnauthorizedOperation: You are not authorized to perform this operation. User: arn:aws:iam::460538899914:user/ci-op-1m6j03lr-ccf9a-minimal-perm is not authorized to perform: ec2:DescribeInstanceTypes because no identity-based policy allows the ec2:DescribeInstanceTypes action
level=error msg=	status code: 403, request id: d64a6b4e-6a3e-43bc-b223-6d2b030c12f7]
Create manifests exit code: 3
```
which is addressed by https://github.com/openshift/installer/pull/9106

r4f4 commented:
/assign @patrickdillon

fahlmant reviewed (commented):
lgtm

r4f4 commented:
Update: rebased on top of current master to fix merge conflicts.

r4f4 commented:
/test e2e-aws-ovn-min-perms

barbacbd reviewed (commented):
/lgtm

patrickdillon commented:
/approve
